### PR TITLE
boards: st: Add support for the Nucleo-H503RB board

### PR DESCRIPTION
**Pull request by SandraArrow**
This PR adds support for nucleo_h503rb board which is based on STM32H503RBT6 SoC. 
 
STM32CubeProgrammer and pyocd are supported as runner. 
The following samples were tested:
- samples/hello_world
- samples/basic/blinky
- samples/basic/button
- samples/subsys/task_wdt

### CODE REVIEW

**Comment by FRASTM** [comment]
Thanks for enabling this nucleo stm32h503 board

File: boards/st/nucleo_h503rb/nucleo_h503rb.dts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+&usart3 {
+	pinctrl-0 = <&usart3_tx_pa4 &usart3_rx_pa3>;
+	pinctrl-names = "default";
+	current-speed = <115200>;
+	status = "okay";
+};
+
+&rng {
+	status = "okay";
```
Comment (FRASTM) [inline boards/st/nucleo_h503rb/nucleo_h503rb.dts:104]: Did you check RNG with related testcase

File: boards/st/nucleo_h503rb/nucleo_h503rb.dts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+};
+
+&iwdg {
+	status = "okay";
+};
+
+&i2c1 {
+	pinctrl-0 = <&i2c1_scl_pb6 &i2c1_sda_pb7>;
```
Comment (FRASTM) [inline boards/st/nucleo_h503rb/nucleo_h503rb.dts:112]: Did you check I2C 1  with related testcase

File: boards/st/nucleo_h503rb/nucleo_h503rb.dts (hunk 1 above)
Comment (SandraArrow) [reply boards/st/nucleo_h503rb/nucleo_h503rb.dts:104]: Yes, I ran the following tests for RNG:
`zephyr/scripts/twister --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200 -p nucleo_h503rb  -T zephyr/tests/subsys/random/rng`

I got the following output:

```shell
INFO    - JOBS: 16 
INFO    - Adding tasks to the queue... 
INFO    - Added initial list of jobs to queue 
INFO    - Total complete:    5/   5  100%  skipped:    1, failed:    0, error:    0 
INFO    - 5 test scenarios (5 test instances) selected, 1 configurations skipped (0 by static filter, 1 at runtime). 
INFO    - 4 of 5 test configurations passed (100.00%), 0 failed, 0 errored, 1 skipped with 0 warnings in 25.50 seconds 
INFO    - In total 4 test cases were executed, 1 skipped on 1 out of total 1 platforms (100.00%) 
INFO    - 4 test configurations executed on platforms, 0 test configurations were only built. 
```

The test case `drivers.rng.random_psa_crypto` was skipped.

File: boards/st/nucleo_h503rb/nucleo_h503rb.dts (hunk 2 above)
Comment (SandraArrow) [reply boards/st/nucleo_h503rb/nucleo_h503rb.dts:112]: No, I had not checked that. Unfortunately, I do not have the necessary hardware for the I2C tests.
I have tested it now with the I2C and EEPROM shell and saw that I have to add a clock-frequency for `i2c1` in the dts file to get it to work.

File: boards/st/nucleo_h503rb/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+Overview
+********
+
+The Nucleo-H503RB board features an ARM |reg| Cortex |reg|-M33 core-based
+STM32H503RBT6 microcontroller with a wide range of connectivity support and
+configurations.
+Here are some highlights of the Nucleo-H503RB board:
+
+- STM32H503RB microcontroller featuring 128 kbytes of Flash memory and 32 Kbytes of
```
Comment (FRASTM) [inline boards/st/nucleo_h503rb/doc/index.rst:14]:
```suggestion
- STM32H503RB microcontroller featuring 128 Kbytes of Flash memory and 32 Kbytes of
```

**Review by FRASTM — COMMENTED** [review]
LGTM

I suggest to consider the https://github.com/zephyrproject-rtos/zephyr/pull/75308   for openOCD runner (depending on the which is merged first )

File: boards/st/nucleo_h503rb/arduino_r3_connector.dtsi [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		gpio-map-pass-thru = <0 0x3f>;
+		gpio-map = <0 0 &gpioa 0 0>,	/* A0 */
+			   <1 0 &gpioa 1 0>,	/* A1 */
+			   <2 0 &gpioa 2 0>,	/* A2 */
+			   <3 0 &gpiob 0 0>,	/* A3 */
+			   <4 0 &gpioc 1 0>,	/* A4 */
+			   <5 0 &gpioc 0 0>,	/* A5 */
+			   <6 0 &gpiob 15 0>,	/* D0 */
+			   <7 0 &gpiob 14 0>,	/* D1 */
+			   <8 0 &gpioa 10 0>,	/* D2 */
```
[the remark this answers is not in the document]
Comment (SandraArrow) [reply boards/st/nucleo_h503rb/arduino_r3_connector.dtsi:21]: I checked it. It is PA10 as described in the user manual for the STM32H5 Nucleo-64 board ([UM3121](https://www.st.com/resource/en/user_manual/um3121-stm32h5-nucleo64-board-mb1814-stmicroelectronics.pdf))

File: boards/st/nucleo_h503rb/doc/index.rst (hunk 3 above)
Comment (SandraArrow) [reply boards/st/nucleo_h503rb/doc/index.rst]: Done

File: boards/st/nucleo_h503rb/board.cmake [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(stm32cubeprogrammer "--erase" "--port=swd" "--reset-mode=hw")
```
[the remark this answers is not in the document]
Comment (SandraArrow) [reply boards/st/nucleo_h503rb/board.cmake]: Done

**Comment by SandraArrow** [comment]
> LGTM
> 
> I suggest to consider the #75308 for openOCD runner (depending on the which is merged first )

I have also adopted the changes from #75308 for this board.

**Review by FRASTM — DISMISSED** [review]

File: boards/st/nucleo_h503rb/board.cmake [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+board_runner_args(openocd "--tcl-port=6666")
+board_runner_args(openocd --cmd-pre-init "gdb_report_data_abort enable")
+board_runner_args(openocd "--no-halt")
+
+board_runner_args(pyocd "--target=stm32h503rbtx")
+
+include(${ZEPHYR_BASE}/boards/common/stm32cubeprogrammer.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
+include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
+# FIXME: official openocd runner not yet available.
```
Comment (marwaiehm-st) [inline boards/st/nucleo_h503rb/board.cmake:14]: To avoid this add an empty line at the end of the file

Comment (SandraArrow) [reply boards/st/nucleo_h503rb/board.cmake]: Done

**Review by marwaiehm-st — APPROVED** [review]
LGTM

**Review by FRASTM — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**